File: agents/javascript/src/tools/list-directory.ts
/**
 * Tool: list_directory
 * Lists directory contents with type and size information.
 * Supports recursive listing with depth control.
 */

import * as fs from 'fs';
import * as path from 'path';

export async function listDirectory(
  args: Record<string, unknown>,
  projectPath: string
): Promise<string> {
  const dirPath = args.path as string ?? '.';
  const recursive = args.recursive as boolean ?? false;
  const maxDepth = args.maxDepth as number ?? 3;

  const fullPath = path.resolve(projectPath, dirPath);

  if (!fs.existsSync(fullPath)) {
    return `Error: Directory not found: ${dirPath}`;
  }

  const stat = fs.statSync(fullPath);
  if (!stat.isDirectory()) {
    return `Error: ${dirPath} is not a directory`;
  }

  const entries: string[] = [];
  listDir(fullPath, projectPath, entries, recursive, maxDepth, 0);

  if (entries.length === 0) {
    return `Directory ${dirPath} is empty`;
  }

  return entries.join('\n');
}

function listDir(
  dir: string,
  projectRoot: string,
  entries: string[],
  recursive: boolean,
  maxDepth: number,
  currentDepth: number
): void {
  let dirEntries: fs.Dirent[];
  try {
    dirEntries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return;
  }

  const indent = '  '.repeat(currentDepth);

  for (const entry of dirEntries) {
    const fullPath = path.join(dir, entry.name);

    if (entry.isDirectory()) {
      // Count children for display
      let childCount = 0;
      try {
        childCount = fs.readdirSync(fullPath).length;
      } catch {
        // ignore
      }
      entries.push(`${indent}[DIR]  ${entry.name}/ (${childCount} items)`);

      if (recursive && currentDepth < maxDepth) {
        listDir(fullPath, projectRoot, entries, recursive, maxDepth, currentDepth + 1);
      }
    } else if (entry.isFile()) {
      try {
        const stat = fs.statSync(fullPath);
        const size = formatSize(stat.size);
        entries.push(`${indent}[FILE] ${entry.name} (${size})`);
      } catch {
        entries.push(`${indent}[FILE] ${entry.name}`);
      }
    }
  }
}

function formatSize(bytes: number): string {
  if (bytes < 1024) return `${bytes}B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)}KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)}MB`;
}

File: agents/python/src/tools/list_directory.py
"""Tool: list_directory — list directory contents with type and size."""

import os
from typing import Any


def list_directory(args: dict[str, Any], project_path: str) -> str:
    dir_path = args.get("path", ".")
    recursive = args.get("recursive", False)
    max_depth = args.get("maxDepth", 3)

    full_path = os.path.normpath(os.path.join(project_path, dir_path))

    if not os.path.exists(full_path):
        return f"Error: Directory not found: {dir_path}"

    if not os.path.isdir(full_path):
        return f"Error: {dir_path} is not a directory"

    entries: list[str] = []
    _list_dir(full_path, project_path, entries, recursive, max_depth, 0)

    if not entries:
        return f"Directory {dir_path} is empty"

    return "\n".join(entries)


def _list_dir(
    dir_path: str,
    project_root: str,
    entries: list[str],
    recursive: bool,
    max_depth: int,
    current_depth: int,
) -> None:
    indent = "  " * current_depth

    try:
        items = sorted(os.listdir(dir_path))
    except PermissionError:
        return

    for item in items:
        full = os.path.join(dir_path, item)

        if os.path.isdir(full):
            try:
                child_count = len(os.listdir(full))
            except PermissionError:
                child_count = 0
            entries.append(f"{indent}[DIR]  {item}/ ({child_count} items)")

            if recursive and current_depth < max_depth:
                _list_dir(full, project_root, entries, recursive, max_depth, current_depth + 1)
        elif os.path.isfile(full):
            try:
                size = os.path.getsize(full)
                entries.append(f"{indent}[FILE] {item} ({_format_size(size)})")
            except OSError:
                entries.append(f"{indent}[FILE] {item}")


def _format_size(bytes_: int) -> str:
    if bytes_ < 1024:
        return f"{bytes_}B"
    if bytes_ < 1024 * 1024:
        return f"{bytes_ / 1024:.1f}KB"
    return f"{bytes_ / (1024 * 1024):.1f}MB"
